### ISSUE TITLE
"Unimatics"

### ISSUE THREAD

**Issue by peterbrittain** [peterbrittain/asciimatics#28]
I've deliberately avoided extended character sets so far as this gets into a whole heap of encoding/decoding pain.  However, it might be that UTF-8 does the trick now, though, and I could move to full unicode.

Support for UTF8 in Linux and OSX looks pretty well established.  However,  Windows support is suspect.  Although there is codepage 65001, it would appear the glyphs are very limited.  More investigation required...

**Comment by peterbrittain** [comment]
In addition to box characters, the new version could also use braille characters for line drawing like https://github.com/yaronn/blessed-contrib

**Comment by peterbrittain** [comment]
Had a little play on Windows/3.x and Linux/2.7.  Looks like output isn't too bad.  I've already got some basic box characters working!  Windows needs work to get up and running - e.g. chcp 65001 plus a custom font install for extended unicode chars - as per http://webservices20.blogspot.co.uk/2015/04/running-terminal-dashboards-on-windows.html

And input is going to be harder...  Curses on Python (up to 3.3) only offers getch() which is now unicode aware, so I'm going to have to do some byte stream parsing before handing off to a codec.  It looks like Windows may need the same unless we pass the unicode char form of key events all the way through.

**Comment by peterbrittain** [comment]
And it's really easy to mess up unicode cross-version compatilibiity in Python.  Sigh.  Make sure to check against http://python-future.org/compatible_idioms.html

**Comment by peterbrittain** [comment]
Turns out Windows input is tricky.  The input for Unicode characters pasted into the console doesnt generate key down events, just key up.  However, standard input only guarantees to get key down in the right order and uses key down for repeats.  In an attempt to be future proof, I've therefore taken to tracking which keys are pressed and just assuming random injected key ups need to be processed as new key presses.  See https://github.com/peterbrittain/asciimatics/commit/58a14c7bdb1b60177f49ff2d4b755ffca2cdb8ae for details.

**Comment by peterbrittain** [comment]
OK - it looks like getch() returns the byte stream, which can be passed to a UTF-8 decoder for curses.  First pass of doing this done in https://github.com/peterbrittain/asciimatics/commit/cea0a94f939769a392b541f4c4dfba2410678449

**Comment by peterbrittain** [comment]
Ok, so we have proved unicode input and output is possible... What's left before closing this?

1. A simple way to set/check whether unicode should be used.
2. Updates to some effects to use better characters when unicode is enabled - box chars, snowflakes, stars, ticks and braille chars for lines.
3. Docs for the new feature - including README and CHANGES

**Comment by norok2** [comment]
It might be worth checking out the `future` package which offers Python3 semantics for Python2 constructs, including some support to the str/unicode.

**Comment by peterbrittain** [comment]
Already using it, but thanks for checking!

**Comment by peterbrittain** [comment]
I've now added a unicode_aware flag for Screen construction, that can be accessed by all other objects to decide whether to display unicode or not.  Next step on that front is to try to provide an auto-detect mode (rather than defaulting to False).

I've also played some more with using unicode chars for slicker output.  Looks like FreeMono isn't great on Windows.  Might be worth finding another...  Perhaps http://dejavu-fonts.org/wiki/Main_Page
